[FSDP()][13/N] Refactor unshard/reshard/grads

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #88123 [FSDP] Rename `unflat_param_name` -> `fqn` for consistency
* #88122 [FSDP] Simplify `_get_buffer_names()`
* #88121 [FSDP] Remove unneeded `torch.no_grad()` context when offloading to CPU
* #88120 [FSDP][Docs] Add note mentioning rate limiter for backward prefetch
* #88040 [FSDP()][27/N] Add forward hook registration
* #87941 [FSDP()][26/N] Move `_lazy_init()` into `_fsdp_root_pre_forward()`
* #87940 [FSDP()][25/N] Add `_post_forward_reshard()`
* #87939 [FSDP()][24/N] Refactor `_lazy_init()`
* #87938 [FSDP()][23/N] Refactor handle attr initialization
* #87937 [FSDP] Simplify `_reset_lazy_init()`
* #87936 [FSDP()][22/N] Refactor `_cast_buffers()` in `_lazy_init()`
* #87935 [FSDP()][21/N] Refactor `_buffer_name_to_orig_dtype` computation
* #87934 [FSDP] Rename `dtype` to `buffer_name_to_dtype`
* #87933 [FSDP] Remove `device` arg from `_cast_buffers()`
* #87932 [FSDP()][20/N][Easy] Move functions in file
* #87931 [FSDP()][18/N] Refactor `pre_forward_unshard()`
* #87930 [FSDP()][17/N] Refactor `_fsdp_root_pre_forward()`
* #87929 [FSDP()][16/N] Refactor post-forward/pre-backward
* #87928 [FSDP()][15/N] Refactor `_init_streams()`
* #87927 [FSDP()][14/N] Refactor pre-forward/post-backward
* **#87926 [FSDP()][13/N] Refactor unshard/reshard/grads**
* #87925 [FSDP()][12/N] Easy cleanup
* #87924 [FSDP()][10/N][11/N] Introduce composable (ctor only)
* #87923 [FSDP()][9/N] Refactor ctor (continued)

This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

Commit:
[FSDP()][13/N] Refactor unshard/reshard/grads

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"




[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

mrshenli reviewed (approved):
Assuming no logic change, just moving code around.

awgu commented on torch/distributed/fsdp/_runtime_utils.py:
```diff
@@ -1,11 +1,96 @@
 from torch.distributed.utils import _to_kwargs
 
 
+@no_type_check
```
Maybe one day when FSDP is pretty stabilized, we can pre-define every attribute properly in `FSDPState.__init__()` and `FullyShardedDataParallel.__init__()`. The current way of just dynamically adding attributes onto `state` is definitely not good. E.g., it is hard to tell what is even defined.

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

Commit:
Update on "[FSDP()][13/N] Refactor unshard/reshard/grads"


This PR is not too complicated. We just move unshard/reshard/grads out to `_runtime_utils.py` and make them take `state: _State` instead of `self`.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
